resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala

```diff
@@ -103,42 +100,43 @@ object SparkKubernetesClientFactory extends Logging {
       .withRequestTimeout(clientType.requestTimeout(sparkConf))
       .withConnectionTimeout(clientType.connectionTimeout(sparkConf))
       .withTrustCerts(sparkConf.get(KUBERNETES_TRUST_CERTIFICATES))
-      .withOption(oauthTokenValue) {
-        (token, configBuilder) => configBuilder.withOauthToken(token)
-      }.withOption(oauthTokenFile) {
-        (file, configBuilder) =>
-            configBuilder.withOauthToken(Files.asCharSource(file, Charsets.UTF_8).read())
-      }.withOption(caCertFile) {
-        (file, configBuilder) => configBuilder.withCaCertFile(file)
-      }.withOption(clientKeyFile) {
-        (file, configBuilder) => configBuilder.withClientKeyFile(file)
-      }.withOption(clientCertFile) {
-        (file, configBuilder) => configBuilder.withClientCertFile(file)
-      }.withOption(namespace) {
-        (ns, configBuilder) => configBuilder.withNamespace(ns)
-      }.build()
-    val factoryWithCustomDispatcher = new OkHttpClientFactory() {
-      override protected def additionalConfig(builder: OkHttpClient.Builder): Unit = {
-        builder.dispatcher(dispatcher)
+      .withOption(oauthTokenValue) { (token, configBuilder) =>
+        configBuilder.withOauthToken(token)
       }
-    }
-    logDebug("Kubernetes client config: " +
-      new ObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(config))
+      .withOption(oauthTokenFile) { (file, configBuilder) =>
+        configBuilder.withOauthToken(Files.asCharSource(file, Charsets.UTF_8).read())
+      }
+      .withOption(caCertFile) { (file, configBuilder) =>
+        configBuilder.withCaCertFile(file)
+      }
+      .withOption(clientKeyFile) { (file, configBuilder) =>
+        configBuilder.withClientKeyFile(file)
+      }
+      .withOption(clientCertFile) { (file, configBuilder) =>
+        configBuilder.withClientCertFile(file)
+      }
+      .withOption(namespace) { (ns, configBuilder) =>
+        configBuilder.withNamespace(ns)
+      }
+      .build()
+    logDebug(
+      "Kubernetes client config: " +
+        new ObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(config))
     new KubernetesClientBuilder()
-      .withHttpClientFactory(factoryWithCustomDispatcher)
```

Comment:
The custom dispatcher uses daemon threads. Not setting this custom dispatcher implies that the we will now have non-daemon threads in the http client.

Reply:
> The custom dispatcher uses daemon threads. 

that's the exact purpose of overwriting the HttpClientFactory

Reply:
Yeah I thought so. How about adding something like this?

```
val clientFactory = HttpClientUtils.getHttpClientFactory
val clientBuilder = new KubernetesClientBuilder().withConfig(config)
clientFactory match {
  case _: OkHttpClientFactory =>
    val dispatcher = new Dispatcher(
      ThreadUtils.newDaemonCachedThreadPool("kubernetes-dispatcher"))
    val factory = new OkHttpClientFactory() {
      override def additionalConfig(builder: OkHttpClient.Builder): Unit = {
        builder.dispatcher(dispatcher)
      }
    }
    clientBuilder.withHttpClientFactory(factory).build()
  case _ => clientBuilder.build()
}
```

Reply:
I'm not against that, after making this change, how do you configure the K8s client builder to use another HTTP client factory without touching the code?

BTW, please restore unnecessary code changes from your patch.

Reply:
In the job jar, you would include a dependency on another http client, for eg.
```
<dependency>
    <groupId>io.fabric8</groupId>
    <artifactId>kubernetes-httpclient-jdk</artifactId>
  </dependency>
```
These dependency jars contain the file `META-INF/services/io.fabric8.kubernetes.client.http.HttpClient$Factory` containing the name of the class implementing the provider, which the K8s client builder should pick up.

The job jar could also include `META-INF/services/io.fabric8.kubernetes.client.http.HttpClient$Factory` with the name of the class we want to use.

I'll restore the changes - somehow running `/dev/scalafmt` introduced those.

Reply:
then how do you choose the exact one you want to use if there are multi impls under the classpath?

Reply:
You can set priority in the provider configuration file:
```
1: <class1>
2: <class2>
```
The code in fabric8 appears to be [honoring](https://github.com/fabric8io/kubernetes-client/blob/main/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/HttpClientUtils.java#L183) that priority.

Reply:
make sense, thanks for sharing the information

Reply:
Sorry @pan3793 minor correction to what I said earlier. The priority comes from the factory itself. So if there are multiple options, we could extend the factory we want and assign priority, like this:
```
class MyHttpClientFactory extends JdkHttpClientFactory {
      override def priority(): Int = -1
}
```
and then add `MyHttpClientFactory` to the configuration file. 

Regardless, with this change it should be possible to choose an implementation without touching code.